src/ppca.jl

```diff
@@ -0,0 +1,192 @@
+                tol::Real=1.0e-6,   # convergence tolerance
+                tot::Integer=1000)  # maximum number of iterations
+
+    check_pcaparams(size(S,1), mean, maxoutdim, 1.)
+
+    d = size(S,1)
+    q = maxoutdim
+    W = eye(d,q)
+    σ² = 0.
+    M⁻¹ = inv(W'W .+ σ²*eye(q))
+
+    i = 1
+    L_old = 0.
+    while i < tot
+        # EM-steps
+        SW = S*W
+        W⁺  = SW*inv(σ²*eye(q) + M⁻¹*W'*SW)
+        σ²⁺ = trace(S - SW*M⁻¹*(W⁺)')/d
+        # new parameters
+        W = W⁺
+        σ² = σ²⁺
+        # log likelihood
+        C = W*W'.+ σ²*eye(d)
+        M⁻¹ = inv(W'*W .+ σ²*eye(q))
+        C⁻¹ = (eye(d) - W*M⁻¹*W')/σ²
+        L = (-n/2)*(log(det(C)) + trace(C⁻¹*S))  # (-n/2)*d*log(2π) omitted
+        # println("$i] ΔL: $(abs(L_old - L)), L: $L")
+        if abs(L_old - L) < tol
+            break
+        end
+        L_old = L
+        i += 1
+    end
+
+    return PPCA(mean, W, σ²)
+end
+
+function bayespca{T<:AbstractFloat}(S::DenseMatrix{T}, mean::Vector{T}, n::Int;
+                 maxoutdim::Int=size(S,1)-1,
+                 tol::Real=1.0e-6,   # convergence tolerance
+                 tot::Integer=1000)  # maximum number of iterations
+
+    check_pcaparams(size(S,1), mean, maxoutdim, 1.)
+
+    d = size(S,1)
+    q = maxoutdim
+    W = eye(d,q) #rand(d,q)
+    wnorm = zeros(q)
+    σ² = 0.
+    M = W'*W .+ σ²*eye(q)
+    M⁻¹ = inv(M)
+    α = zeros(q)
+
+    i = 1
+    L_old = 0.
+    while i < tot
+        # EM-steps
+        SW = S*W
+        W⁺  = SW*inv(σ²*(eye(q)+diagm(α)*M/n) + M⁻¹*W'*SW)
+        σ²⁺ = trace(S - SW*M⁻¹*(W⁺)')/d
+        # new parameters
+        W = W⁺
+        σ² = σ²⁺
+        @inbounds for j in 1:q
+            wnorm[j] = norm(W[:,j])^2
+            α[j] = wnorm[j] < eps() ? maxintfloat(Float64) : d/wnorm[j]
+        end
+        # log likelihood
+        C = W*W'.+ σ²*eye(d)
+        M = W'*W .+ σ²*eye(q)
+        M⁻¹ = inv(M)
+        C⁻¹ = (eye(d) - W*M⁻¹*W')/σ²
+        L = (-n/2)*(log(det(C)) + trace(C⁻¹*S))  # (-n/2)*d*log(2π) omitted
+        # println("$i] ΔL: $(abs(L_old - L)), L: $L")
+        if abs(L_old - L) < tol
+            break
+        end
+        L_old = L
+        i += 1
+    end
+
+    return PPCA(mean, W[:,wnorm .> 0.], σ²)
+end
+
+## interface functions
+
+function fit{T<:AbstractFloat}(::Type{PPCA}, X::DenseMatrix{T};
+             method::Symbol=:ml,
+             maxoutdim::Int=size(X,1)-1,
+             mean=nothing,
+             tol::Real=1.0e-6,   # convergence tolerance
+             tot::Integer=1000)  # maximum number of iterations
+
+    d, n = size(X)
+
+    # process mean
+    mv = preprocess_mean(X, mean)
+    (isempty(mv) || length(mv) == d) ||
+            throw(DimensionMismatch("Dimensions of weight matrix and mean are inconsistent."))
+
+    if method == :ml
+        Z = centralize(X, mv)
+        M = ppcaml(Z, mv, maxoutdim=maxoutdim, tol=tol)
+    elseif method == :em
+        S = Base.covm(X, isempty(mv) ? 0 : mv, 2)
```

Comment:
I guess Julia 0.4 doesn't like this signature for `covm`. I'd be fine dropping 0.4 support at this point though.